python_modules/dagster/dagster/_core/decorator_utils.py

```diff
@@ -128,3 +147,151 @@ def has_at_least_one_parameter(
+    we need to create a new `property` object with the updated `fget` function that preserves the
+    rest of the property's attributes. We also want to make sure we copy over any special attributes
+    we stored on the `fget` (since we can't and therefore don't store them on the `property`
+     object).
+    """
+    from dagster._annotations import copy_annotations
+
+    curr_fn = get_decorator_target(decoratable)
+    new_fn = functools.update_wrapper(new_fn, curr_fn)
+    # from types import FunctionType
+    # if inspect.isfunction(decoratable) or isinstance(decoratable, FunctionType):
+    #     val = new_fn
+    if isinstance(decoratable, property):
+        val = property(new_fn, decoratable.fset, decoratable.fdel, decoratable.__doc__)
+        copy_annotations(val, decoratable)
+    elif isinstance(decoratable, staticmethod):
+        val = staticmethod(new_fn)
+        copy_annotations(val, decoratable)
+    elif isinstance(decoratable, classmethod):
+        val = classmethod(new_fn)
+        copy_annotations(val, decoratable)
+    elif isinstance(decoratable, type):
+        constructor_name = "__new__" if issubclass(decoratable, tuple) else "__init__"
+        setattr(decoratable, constructor_name, new_fn)
+        val = decoratable
+    elif is_resource_def(decoratable):
+        setattr(decoratable, "_resource_fn", new_fn)
+        val = decoratable
+    elif is_op_def(decoratable):
+        if is_decorated_op_function(decoratable.compute_fn):
+            setattr(decoratable.compute_fn, "decorated_fn", new_fn)
+        else:
+            setattr(decoratable, "compute_fn", new_fn)
+        val = decoratable
+    else:  # plain function
+        val = new_fn
+
+    # Pyright is unable to infer based on our checks that the return type is the same as the input
+    # type.
+    return cast(T_Decoratable, val)
+
+
+def is_decoratable(obj: Any) -> TypeGuard[Decoratable]:
+    """Return `True` if the passed object is a decoratable object."""
+    return (
+        isinstance(obj, type)
+        or isinstance(obj, (property, staticmethod, classmethod))
+        or inspect.isfunction(obj)
+        or is_resource_def(obj)
+        or is_op_def(obj)
+    )
+
+
+def is_resource_def(obj: Any) -> TypeGuard["ResourceDefinition"]:
+    """Use this in place of `isinstance` when `ResourceDefinition`
+    can't be imported yet (due to circular import issues).
+    """
+    return hasattr(obj, "_resource_fn")
+
+
+def is_op_def(obj: Any) -> TypeGuard["OpDefinition"]:
+    """Use this in place of `isinstance` when `OpDefinition`
+    can't be imported yet (due to circular import issues).
+    """
+    return hasattr(obj, "_compute_fn")
+
+
+def is_decorated_op_function(
+    obj: Union[Callable, "DecoratedOpFunction"]
+) -> TypeGuard["DecoratedOpFunction"]:
+    """Use this in place of `isinstance` when `DecoratedOpFunction`
+    can't be imported yet (due to circular import issues).
+    """
+    return hasattr(obj, "decorated_fn")
```

Comment:
more yikes, these are quite spicy

Reply:
can we confirm against the `__name__` or something too? Should at least be under direct explicit test since someone refactoring implementations can break this

Reply:
Yeah these are rough but this is a tricky situation. All the annotation code is being executed at import time, which means that if we used e.g `isinstance(obj, ResourceDefinition)`, it would get executed the first time an annotation gets encountered, which is before `ResourceDefinition`. I tried a local import of `ResourceDefinition` but it led to a tangled mess of circular import errors.

While I recognize this is ugly and not an approach I would use in a more general context, these functions are _only_ called when we've decorated something with one of these special decorators, and IMO it is OK in this limited context.

Reply:
I updated the implementation to check the `mro` of `obj.__class__` for `"ResourceDefinition"`